common/src/main/java/com/wynntils/gui/render/MapRenderer.java

```diff
@@ -99,10 +116,226 @@ public static void renderMapQuad(
+                    servicePoi);
+        }
+
+        if (McUtils.mc().screen instanceof MainMapScreen mainMapScreen) {
+            mainMapScreen.setHovered(hovered);
+        }
+
+        if (hovered != null && mouseCoordinates != null) {
+            float textureXPosition = hovered.getRenderX(mapCenterX, centerX, 1f / scale);
+            float textureZPosition = hovered.getRenderZ(mapCenterZ, centerZ, 1f / scale);
+
+            float width = hovered.getIcon().width() * poiScale;
+            float height = hovered.getIcon().height() * poiScale;
+
+            poseStack.pushPose();
+
+            float renderX = textureXPosition - width / 2f;
+            float renderZ = textureZPosition - height / 2f;
+            poseStack.translate(renderX, renderZ, 0);
+
+            FontRenderer.getInstance()
+                    .renderText(
+                            poseStack,
+                            hovered.getName(),
+                            width / 2f,
+                            20,
+                            CommonColors.GREEN,
+                            HorizontalAlignment.Center,
+                            VerticalAlignment.Top,
+                            FontRenderer.TextShadow.OUTLINE);
+
+            poseStack.popPose();
+        }
+    }
+
+    public static void renderServicePoi(
+            PoseStack poseStack,
+            Pair<Integer, Integer> mouseCoordinates,
+            float mapCenterX,
+            float mapCenterZ,
+            float centerX,
+            float centerZ,
+            boolean followPlayerRotation,
+            float scale,
+            float poiScale,
+            ServicePoi servicePoi) {
+        // TODO: This is really basic at the moment
+        //       Add fading, and other configs
+
+        float textureXPosition;
+        float textureZPosition;
+
+        if (followPlayerRotation) {
+            final double rotationRadians = Math.toRadians(McUtils.player().getYRot());
+            final float sinRotationRadians = (float) StrictMath.sin(rotationRadians);
+            final float cosRotationRadians = (float) -StrictMath.cos(rotationRadians);
+
+            float dX = (servicePoi.getLocation().getX() - mapCenterX) / scale;
+            float dZ = (servicePoi.getLocation().getZ() - mapCenterZ) / scale;
+
+            textureXPosition = centerX + (dX * cosRotationRadians - dZ * sinRotationRadians);
+            textureZPosition = centerZ + (dX * sinRotationRadians + dZ * cosRotationRadians);
+        } else {
+            textureXPosition = servicePoi.getRenderX(mapCenterX, centerX, 1f / scale);
+            textureZPosition = servicePoi.getRenderZ(mapCenterZ, centerZ, 1f / scale);
+        }
+
+        float width = servicePoi.getIcon().width() * poiScale;
+        float height = servicePoi.getIcon().height() * poiScale;
+
+        float renderX = textureXPosition - width / 2f;
+        float renderZ = textureZPosition - height / 2f;
+
+        if (mouseCoordinates != null) {
```

Comment:
Maybe you can extract this out of renderServicePoi and do it separately when looping over all service POIs. If you do it first, you can send in a "isHovered" or whatever to say that the icon should be bigger.

Reply:
Well, I would have to extract a lot of math and calculate everything twice. Not good.

Reply:
All of this
```java
        float textureXPosition;
        float textureZPosition;

        if (followPlayerRotation) {
            final double rotationRadians = Math.toRadians(McUtils.player().getYRot());
            final float sinRotationRadians = (float) StrictMath.sin(rotationRadians);
            final float cosRotationRadians = (float) -StrictMath.cos(rotationRadians);

            float dX = (servicePoi.getLocation().getX() - mapCenterX) / scale;
            float dZ = (servicePoi.getLocation().getZ() - mapCenterZ) / scale;

            textureXPosition = centerX + (dX * cosRotationRadians - dZ * sinRotationRadians);
            textureZPosition = centerZ + (dX * sinRotationRadians + dZ * cosRotationRadians);
        } else {
            textureXPosition = getRenderX(servicePoi, mapCenterX, centerX, 1f / scale);
            textureZPosition = getRenderZ(servicePoi, mapCenterZ, centerZ, 1f / scale);
        }

        float width = servicePoi.getIcon().width() * poiScale;
        float height = servicePoi.getIcon().height() * poiScale;

        float renderX = textureXPosition - width / 2f;
        float renderZ = textureZPosition - height / 2f;
```

Reply:
Oh, I missed that. Hm.

Reply:
I have a suggestion, but I understand if you do not want to follow it. Turn this calculation into a Pair<float,float> getRenderPosition() that returns the renderX and renderZ, to the loop. Then use those to a) check the hover and b) call drawScalingTexturedRect.

I think it would be better, by still containing the maths separately, but showing that we are doing two things (hover calucation and rendering) per poi.